staker/staker.gno

```diff
@@ -71,310 +86,142 @@ func StakeToken(tokenId uint64) (string, string, string) {
 		CalcPoolPosition()
 	}
 
-	// check whether tokenId already staked or not
-	_, exist := deposits[tokenId]
-	if exist {
-		panic(addDetailToError(
-			errAlreadyStaked,
-			ufmt.Sprintf("staker.gno__StakeToken() || tokenId(%d) already staked", tokenId),
-		))
-	}
-
 	owner := gnft.OwnerOf(tid(tokenId))
-
-	// if caller is owner
 	caller := std.PrevRealm().Addr()
-	callerIsOwner := owner == caller
 
-	// stakerIsOwner
-	stakerIsOwner := owner == consts.STAKER_ADDR
-
-	if !(callerIsOwner || stakerIsOwner) {
-		panic(addDetailToError(
-			errNoPermission,
-			ufmt.Sprintf("staker.gno__StakeToken() || caller(%s) or staker(%s) is not owner(%s) of tokenId(%d)", caller, consts.STAKER_ADDR, owner, tokenId),
-		))
-	}
-
-	// check pool path from tokenid
-	poolPath := pn.PositionGetPositionPoolKey(tokenId)
-
-	// check if target pool doesn't have internal or external incentive then panic
-	hasInternal := poolHasInternal(poolPath)
-	hasExternal := poolHasExternal(poolPath)
-	if hasInternal == false && hasExternal == false {
-		panic(addDetailToError(
-			errNonIncentivizedPool,
-			ufmt.Sprintf("staker.gno__StakeToken() || can not stake position to non incentivized pool(%s)", poolPath),
-		))
-	}
-
-	// check tokenId has liquidity or not
-	liqStr := pn.PositionGetPositionLiquidityStr(tokenId)
-	liquidity := u256.MustFromDecimal(liqStr)
-	if liquidity.Lte(u256.Zero()) {
-		panic(addDetailToError(
-			errZeroLiquidity,
-			ufmt.Sprintf("staker.gno__StakeToken() || tokenId(%d) has no liquidity", tokenId),
-		))
+	result, err := calculateStakeData(tokenId, owner, caller)
+	if err != nil {
+		panic(err)
 	}
 
-	// staked status
-	deposit := deposits[tokenId]
-	deposit.owner = std.PrevRealm().Addr()
-	deposit.numberOfStakes++
-	deposit.stakeTimestamp = time.Now().Unix()
-	deposit.stakeHeight = std.GetHeight()
-	deposit.targetPoolPath = poolPath
-	deposits[tokenId] = deposit
-
-	if callerIsOwner { // if caller is owner, transfer NFT ownership to staker contract
-		transferDeposit(tokenId, consts.STAKER_ADDR)
+	if err := applyStake(result); err != nil {
+		panic(err)
```

Comment:
no need err codes?

Reply:
The `applyStake` already returns `errNoPermission` in the `transferDeposit` function, so it does not have to handle it again.